### PR TITLE
Propagate marked attributes from method type parameters to type parameters of synthesized types.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to https://github.com/dotnet/roslyn/issues/73920.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 1]
```diff
@@ -8553,5 +8553,40 @@ public void ParamsArray_ThisModifier_02()
                 // var lam = (params this int[] xs) => xs.Length;
                 Diagnostic(ErrorCode.ERR_ThisInBadContext, "this").WithLocation(1, 19));
         }
+
+        [Fact]
+        public void CompilerLoweringPreserveAttribute_01()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:8558]: Is it worth adding some tests for lambdas observed as expression trees, to make sure there's nothing different about how we do the machinery lowering for that scenario?

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:8558]: > Is it worth adding some tests for lambdas observed as expression trees, to make sure there's nothing different about how we do the machinery lowering for that scenario?

I do not think so.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs [hunk 2]
```diff
@@ -602,32 +602,71 @@ private TypeParameterConstraintKind GetConstraintKinds()
         }
     }
 
-    internal sealed class SourceMethodTypeParameterSymbol : SourceTypeParameterSymbolBase
+    internal abstract class SourceMethodTypeParameterSymbol : SourceTypeParameterSymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:605]: >abstract class SourceMethodTypeParameterSymbol [](http://example.com/codeflow?start=13&length=46)

What does this refactoring enable?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs [hunk 3]
```diff
@@ -34,6 +34,19 @@ public override bool IsImplicitlyDeclared
 
         internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, ref ArrayBuilder<CSharpAttributeData> attributes)
         {
+            if (ContainingSymbol.Kind == SymbolKind.NamedType &&
+                _underlyingTypeParameter.OriginalDefinition is SourceMethodTypeParameterSymbol &&
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs:38]: >is SourceMethodTypeParameterSymbol [](http://example.com/codeflow?start=60&length=34)

Consider capturing in a local and reusing below: `is SourceMethodTypeParameterSymbol underlyingDefinition`

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:605]: > What does this refactoring enable?

All implementations of method type parameters in source have a common base distinct from the base of type type parameters

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs:38]: > > is SourceMethodTypeParameterSymbol
> 
> Consider capturing in a local and reusing below: `is SourceMethodTypeParameterSymbol underlyingDefinition`

I'll make this adjustment in the next PR

**Pull request merged**